### PR TITLE
feat: Headless readiness

### PR DESCRIPTION
**Pull request by fsbraun**
## Description

Regular projects distribute their content through serving pages at their corresponding URL as defined in the page tree. This is achieved by adding `path("", include(cms.urls))` to a project's `urls.py`.  Also, the CMS requires `CMS_TEMPLATES` or `CMS_TEMPLATE_DIR` to be set.

This PR allows setting up **headless-ready projects** by:

1. Not requiring `path("", include(cms.urls))` being part of the projects URLs. All editing is done in the admin and the edit, structure and preview endpoints.

   The structure endpoint can be used on read-only objects (such as published pages), too, allowing to view the tree and copy plugins. All modifying actions are disabled for read-only objects. 

2. Not requiring any templates to be defined.  If no templates are available, all interaction runs through the structure endpoint. While the editing experience might be limited, this can be a fast way of setting up a project. 

3. This PR moves the create wizard URL to the placeholder admin (which contains all editing endpoints) (but also keeps the previous endpoint in `cms.urls` for the wizard for compatibility reasons).

4. The data bridge allows the frontend to update the structure board after changes in the plugin tree. This PR changes the HTML response for changes in the plugin tree from inlining the data in a script to a script tag containing JSON objects which are read and passed to the data bridge.

## Related resources

* #7835 
* #7927

## Checklist

* [x] I have opened this pull request against ``develop-4``
* [x] I have added or modified the tests when changing logic
* [x] I have followed [the conventional commits guidelines](https://www.conventionalcommits.org/) to add meaningful information into the changelog
* [x] I have read the [contribution guidelines ](https://github.com/django-cms/django-cms/blob/develop/CONTRIBUTING.rst) and I have joined #workgroup-pr-review on [Slack](https://www.django-cms.org/slack) to find a “pr review buddy” who is going to review my pull request.

### CODE REVIEW

**Comment by vinitkumar** [comment]
@fsbraun Great work on this. I would love to see how this works in practise and what we can do with it?

**Comment by fsbraun** [comment]
This PR includes JS and CSS changes to disable UI items in the structure board. Those changes only become visible after running `gulp sass bundle`.

Still, a headless-ready project for testing can be quickly set up using these steps:
* Create a fresh venv: `python3 -m venv .venv`
* Activate it: `source .venv/bin/activate`
* Install PR: `pip install git+https://github.com/fsbraun/django-cms@feat/data-bridge`
* Create project: `djangocms headlesstest --template https://github.com/django-cms/cms-template/tarball/feat/headless`
* Spin up test server: `cd headlesstest && ./manage.py runserver`
* Go to admin in the browser: `http://127.0.0.1/en/admin/`

The template only installs djangocms-text as plugins but works with any plugins.

@vinitkumar @marksweb

File: docs/how_to/21-headless.rst [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+
+.. code-block:: python
+
+    CMS_PLACEHOLDERS = (
+        ('single', ('content'), _('Single placeholder')),
+        ('two_column', ('left', 'right'), _('Two columns')),
+    )
+
+The ``CMS_PLACEHOLDERS`` setting is a list of tuples. Each tuple represents a
+placeholder configuration. The first element of the tuple is the name of the
```
Comment (macolo) [inline docs/how_to/21-headless.rst:106]: I suggest we link to `CMS_PLACEHOLDER_CONF` (https://docs.django-cms.org/en/latest/reference/configuration.html#cms-placeholder-conf) here, it wasn't intuitive to me at first glance what "placeholder configuration" means.

File: docs/how_to/21-headless.rst [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+placeholder configuration. The second element is a tuple of placeholder names.
```
Comment (macolo) [inline docs/how_to/21-headless.rst:107]: It's not clear to me why a placeholder can have multiple names.

File: docs/how_to/21-headless.rst [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+The third element is the verbose description of the placeholder configuration
+which will be shown in the user interface.
```
Comment (macolo) [inline docs/how_to/21-headless.rst:109]: Is this referring to the editor's structure sidebar?

File: docs/reference/configuration.rst [hunk 4]
```diff
@@ -168,6 +164,23 @@ for translation.
...
+CMS_PLACEHOLDERS
+================
+
+default
+    ``(('', ('content',), _("Single placeholder")),)``
+
+A list of placeholders that can be added to a page. The first element of the
+tuple is the name of the placeholder configuration. The second element is a
+tuple of placeholder names. The third element is the verbose description of the
+placeholder configuration which will be shown in the user interface.ed
```
Comment (macolo) [inline docs/reference/configuration.rst:178]: Typo at the end of the line?

**Comment by fsbraun** [comment]
@macolo I've tried to improve the clarity of the `CMS_PLACEHOLDERS` setting. It is supposed to replace `CMS_TEMPLATES` in a headless-only config.

File: cms/admin/pageadmin.py [hunk 5]
```diff
@@ -1194,8 +1194,13 @@ def change_template(self, request, object_id):
 
         to_template = request.POST.get("template", None)
```
Comment (marksweb) [inline cms/admin/pageadmin.py:1195]: What does "to" mean in this context? I'm trying to make sense of the variable naming.

Is it something like the page content is going to go in to this template?

Comment (fsbraun) [reply cms/admin/pageadmin.py:1195]: @marksweb Good question. I assume it should signify the value that goes into the `template` of the page content object. This, of course, is the template that the page content will be rendered in. 

A more traditional way of naming would just be `template`.

**Comment by macolo** [comment]
@fsbraun the docs now make a lot of sense to me, easy to understand! 💪  I like the design of this functionality, it seems quite intuitive to me from a developer's perspective. 👍

**Review by macolo — APPROVED** [review]

**Pull request merged**